core/src/main/java/it/pagopa/selfcare/external_api/core/InstitutionServiceImpl.java

```diff
@@ -75,90 +74,47 @@ public List<Product> getInstitutionUserProductsV2(String institutionId, String u
+                    userProduct.setRole(evaluateRoles(userInstitution));
+                    userProduct.setRoles(userInstitution.getProducts().stream().map(OnboardedProductResponse::getProductRole).toList());
+                    return userProduct;
+                })
+                .collect(Collectors.toList());
+
+        log.debug(LogUtils.CONFIDENTIAL_MARKER, "getInstitutionProductUsers result = {}", userProductResponses);
         log.trace("getInstitutionProductUsers end");
-        return userInfos;
+        return userProductResponses;
     }
 
-    private UserInfo toUserInfos(UserInstitution userInstitution, OnboardedProductResponse onboardedProduct) {
-        SelfCareAuthority selfCareAuthority = Arrays.stream(PartyRole.values())
-                    .filter(partyRole -> partyRole.name().equals(onboardedProduct.getRole()))
-                    .findAny()
-                    .map(PartyRole::getSelfCareAuthority)
-                    .orElseThrow(() -> new RuntimeException(String.format("Role for user %s and institution %s not found!", userInstitution.getUserId(), userInstitution.getInstitutionId())));
-
-        UserInfo userInfo = new UserInfo();
-        userInfo.setId(userInstitution.getUserId());
-        userInfo.setStatus(onboardedProduct.getStatus());
-        userInfo.setRole(selfCareAuthority);
-        userInfo.setPartyRole(PartyRole.valueOf(onboardedProduct.getRole()));
-        ProductInfo productInfo = new ProductInfo();
-        productInfo.setId(onboardedProduct.getProductId());
-        RoleInfo roleInfo = new RoleInfo();
-        roleInfo.setRelationshipId(userInstitution.getUserId());
-        roleInfo.setSelcRole(selfCareAuthority);
-        roleInfo.setRole(onboardedProduct.getProductRole());
-        roleInfo.setStatus(onboardedProduct.getStatus());
-        ArrayList<RoleInfo> roleInfos = new ArrayList<>();
-        roleInfos.add(roleInfo);
-        productInfo.setRoleInfos(roleInfos);
-        Map<String, ProductInfo> products = new HashMap<>();
-        products.put(productInfo.getId(), productInfo);
-        userInfo.setProducts(products);
-        userInfo.setInstitutionId(userInstitution.getInstitutionId());
-        userInfo.setUserUuidMail(userInstitution.getUserMailUuid());
-        return userInfo;
-    }
+    private PartyRole evaluateRoles(UserInstitution userInstitution) {
+        List<PartyRole> roles = userInstitution.getProducts().stream()
+                .map(OnboardedProductResponse::getRole)
+                .filter(role -> Arrays.stream(PartyRole.values()).anyMatch(partyRole -> partyRole.name().equals(role)))
+                .map(PartyRole::valueOf)
+                .toList();
 
-    private static final BinaryOperator<UserInfo> USER_INFO_MERGE_FUNCTION = (userInfo1, userInfo2) -> {
-        String id = userInfo2.getProducts().keySet().toArray()[0].toString();
+        return Collections.min(roles);
```

Comment:
you could handle when `userInstitution.getProducts()` or `roles` is empty, it should return null value